### PR TITLE
WIP DO NOT MERGE - Plat 126/define hid device

### PR DESCRIPTION
**Pull request by KeepKeyBrett**
add HIDDevice class to allow users to talk to a HID device in the node environment

### CODE REVIEW

File: src/HIDDevice.ts [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+
+  protected async read (): Promise<ByteBuffer> {
+    let first = this.bufferQueue[0]
+    if (!first) throw new Error('Queue is empty')
+    // Check that buffer starts with: "?##" [ 0x3f, 0x23, 0x23 ]
+    // "?" = USB marker, "##" = KeepKey magic bytes
+    // Message ID is bytes 4-5. Message length starts at byte 6.
+    const valid = (first.getUint32(0) & 0xffffff00) === 0x3f232300
+    const msgLength = first.getUint32(5)
+    if (valid && msgLength >= 0 && msgLength < 131072) { // 128KB max message size
```
Comment (elmutt) [inline src/HIDDevice.ts]: maybe define 131072 as a constant as 1024 * 128 to be more clear?

File: src/HIDDevice.ts [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+      return new ByteBuffer(0)
+    }
+  }
+
+  protected async write (buffer: ByteBuffer): Promise<void> {
+    for (let i = 0; i < buffer.limit; i += SEGMENT_SIZE) {
+      let segment = buffer.toArrayBuffer().slice(i, i + SEGMENT_SIZE)
+      let padding = new Array(SEGMENT_SIZE - segment.byteLength + 1).join('\0')
+      let fragments: Array<any> = []
+      fragments.push([63])
```
Comment (elmutt) [inline src/HIDDevice.ts]: is this SEGMENT_SIZE?

**Comment by elmutt** [comment]
Might be cool to have comments with examples of the data structures that are being manipulated so its clear why certain elements are being sliced and manipulated, etc

File: src/HIDDevice.ts [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+  }
+
+  public async disconnect (): Promise<void> {
+    console.log('invoked disconnect')
+    return this.hidDevice.close()
+  }
+
+  public getEntropy (length: number): Uint8Array {
+    console.log('invoked getEntropy')
+    return new Uint8Array(length)
```
Comment (spiceboi) [inline src/HIDDevice.ts]: https://nodejs.org/docs/latest/api/crypto.html#crypto_crypto_randombytes_size_callback

Comment (keepkeyjon) [reply src/HIDDevice.ts]: yeah, empty entropy is not cool

File: src/HIDDevice.ts (hunk 2 above)
Comment (keepkeyjon) [reply src/HIDDevice.ts]: yes, should be.

File: src/HIDDevice.ts [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+        buffer[k] = first.getUint8(k)
+      }
+      let offset = first.byteLength
+
+      let currentBufferIndex = 1
+
+      // this is where things are going to shit
+      while (currentBufferIndex < this.bufferQueue.length) {
+        const next = this.bufferQueue[currentBufferIndex]
+        // Drop USB "?" packet identifier in the first byte
```
Comment (keepkeyjon) [inline src/HIDDevice.ts]: s/packet identifier/USB reportId/

File: src/HIDDevice.ts [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+    const sleep = () => new Promise(resolve => setTimeout(resolve, 2000))
+    await sleep()
+    return this.read()
+  }
+
+  protected async read (): Promise<ByteBuffer> {
+    let first = this.bufferQueue[0]
+    if (!first) throw new Error('Queue is empty')
+    // Check that buffer starts with: "?##" [ 0x3f, 0x23, 0x23 ]
+    // "?" = USB marker, "##" = KeepKey magic bytes
```
Comment (keepkeyjon) [inline src/HIDDevice.ts]: s/USB marker/USB reportId/

File: src/HIDDevice.ts [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+  }
+
+  public getEntropy (length: number): Uint8Array {
+    console.log('invoked getEntropy')
+    return new Uint8Array(length)
+  }
+
+  public async sendRaw (buffer: ByteBuffer): Promise<ByteBuffer> {
+    await this.write(buffer)
+    // need to check for completed message instead of sleeping
```
Comment (keepkeyjon) [inline src/HIDDevice.ts]: is this a TODO?

**Pull request closed without merging**